File: library/src/androidTest/java/com/doctoror/particlesdrawable/ParticlesSceneTest.java
/*
 * Copyright (C) 2017 Yaroslav Mytkalyk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.doctoror.particlesdrawable;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * {@link ParticlesScene} test
 */
public final class ParticlesSceneTest {

    @Test
    public void testPointsEmptyByDefault() {
        final ParticlesScene scene = new ParticlesScene();
        assertTrue(scene.getMutablePoints().isEmpty());
    }

    @Test
    public void testAddTwoPoints() {
        final ParticlesScene scene = new ParticlesScene();

        final Particle p1 = new Particle();
        final Particle p2 = new Particle();

        scene.addParticleData(p1);
        scene.addParticleData(p2);

        assertFalse(scene.getMutablePoints().isEmpty());

        assertTrue(scene.getMutablePoints().contains(p1));
        assertTrue(scene.getMutablePoints().contains(p2));
    }

    @Test
    public void testRemoveFirstPointWhenEmpty() {
        new ParticlesScene().removeFirstParticle();
    }

    @Test
    public void testRemoveFirstPointWithTwoPoints() {
        final ParticlesScene scene = new ParticlesScene();

        final Particle p1 = new Particle();
        final Particle p2 = new Particle();

        scene.addParticleData(p1);
        scene.addParticleData(p2);

        scene.removeFirstParticle();

        assertFalse(scene.getMutablePoints().contains(p1));
        assertTrue(scene.getMutablePoints().contains(p2));

        scene.removeFirstParticle();

        assertFalse(scene.getMutablePoints().contains(p2));
    }

    @Test
    public void testClearPointsWithTwoPoints() {
        final ParticlesScene scene = new ParticlesScene();

        final Particle p1 = new Particle();
        final Particle p2 = new Particle();

        scene.addParticleData(p1);
        scene.addParticleData(p2);

        scene.clearParticles();
        assertTrue(scene.getMutablePoints().isEmpty());
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetFrameDelayNegative() {
        new ParticlesScene().setFrameDelay(-1);
    }

    @Test
    public void testSetFrameDelay() {
        final ParticlesScene s = new ParticlesScene();
        s.setFrameDelay(1);
        assertEquals(1, s.getFrameDelay());
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testStepMultiplierNegative() {
        new ParticlesScene().setStepMultiplier(-0.01f);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testStepMultiplierNAN() {
        new ParticlesScene().setStepMultiplier(Float.NaN);
    }

    @Test
    public void testSetStepMultiplier() {
        final ParticlesScene s = new ParticlesScene();
        s.setStepMultiplier(0f);
        assertEquals(0, s.getStepMultiplier(), Config.ASSERT_DELTA);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeInvalidFirstArgument() {
        new ParticlesScene().setDotRadiusRange(0.49f, 1f);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeInvalidSecondArgument() {
        new ParticlesScene().setDotRadiusRange(2f, 0f);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeInvalidBothArguments() {
        new ParticlesScene().setDotRadiusRange(0.1f, -2f);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeFirstArgumentNAN() {
        new ParticlesScene().setDotRadiusRange(Float.NaN, 1f);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeSecondArgumentNAN() {
        new ParticlesScene().setDotRadiusRange(1f, Float.NaN);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeBothArgumentsNAN() {
        new ParticlesScene().setDotRadiusRange(Float.NaN, Float.NaN);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testSetDotRadiusRangeMaxLessThanMin() {
        new ParticlesScene().setDotRadiusRange(0.7f, 0.6f);
    }

    @Test
    public void testSetDotRadiusRange() {
        final ParticlesScene s = new ParticlesScene();
        s.setDotRadiusRange(0.5f, 0.6f);
        assertEquals(0.5, s.getMinDotRadius(), Config.ASSERT_DELTA);
        assertEquals(0.6, s.getMaxDotRadius(), Config.ASSERT_DELTA);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetLineThicknessInvalidArgument() {
        new ParticlesScene().setLineThickness(0.99f);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetLineThicknessNAN() {
        new ParticlesScene().setLineThickness(Float.NaN);
    }

    @Test
    public void testSetLineThickness() {
        final ParticlesScene s = new ParticlesScene();
        s.setLineThickness(1f);
        assertEquals(1, s.getLineThickness(), Config.ASSERT_DELTA);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetLineDistanceInvalidArgument() {
        new ParticlesScene().setLineDistance(Float.NEGATIVE_INFINITY);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetLineDistanceNAN() {
        new ParticlesScene().setLineDistance(Float.NaN);
    }

    @Test
    public void testSetLineDistance() {
        final ParticlesScene s = new ParticlesScene();
        s.setLineDistance(0f);
        assertEquals(0, s.getLineDistance(), Config.ASSERT_DELTA);
    }

    @SuppressWarnings("Range")
    @Test(expected = IllegalArgumentException.class)
    public void testSetNumDotsInvalidArgument() {
        new ParticlesScene().setNumDots(-1);
    }

    @Test
    public void testSetNumDots() {
        final ParticlesScene s = new ParticlesScene();
        s.setNumDots(0);
        assertEquals(0, s.getNumDots());
    }

    @Test
    public void testSetLineColor() {
        final ParticlesScene s = new ParticlesScene();
        s.setLineColor(2);
        assertEquals(2, s.getLineColor());
    }
}

File: library/src/test/java/com/doctoror/particlesdrawable/ParticlesSceneTest.kt
/*
 * Copyright (C) 2017 Yaroslav Mytkalyk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.doctoror.particlesdrawable

import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@Config(manifest = Config.NONE)
@RunWith(RobolectricTestRunner::class)
class ParticlesSceneTest {

    private val underTest = ParticlesScene()

    @Test
    fun addTwoParticles() {
        // When
        underTest.setParticleData(
                0,
                1f,
                2f,
                3f,
                4f,
                5f,
                6f
        )

        underTest.setParticleData(
                1,
                7f,
                8f,
                9f,
                10f,
                11f,
                12f
        )

        // Then
        assertEquals(1f, underTest.getParticleX(0))
        assertEquals(2f, underTest.getParticleY(0))
        assertEquals(3f, underTest.getParticleDirectionCos(0))
        assertEquals(4f, underTest.getParticleDirectionSin(0))
        assertEquals(5f, underTest.radiuses.get(0))
        assertEquals(6f, underTest.getParticleStepMultiplier(0))

        assertEquals(7f, underTest.getParticleX(1))
        assertEquals(8f, underTest.getParticleY(1))
        assertEquals(9f, underTest.getParticleDirectionCos(1))
        assertEquals(10f, underTest.getParticleDirectionSin(1))
        assertEquals(11f, underTest.radiuses.get(1))
        assertEquals(12f, underTest.getParticleStepMultiplier(1))
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetFrameDelayToNegative() {
        underTest.frameDelay = -1
    }

    @Test
    fun setsFrameDelay() {
        underTest.frameDelay = 1
        assertEquals(1, underTest.frameDelay.toLong())
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetStepMultiplierToNegative() {
        underTest.stepMultiplier = -0.01f
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetMultiplierNAN() {
        underTest.stepMultiplier = java.lang.Float.NaN
    }

    @Test
    fun setsStepMultiplier() {
        underTest.stepMultiplier = 1f
        assertEquals(1f, underTest.stepMultiplier, ASSERT_DELTA)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeInvalidFirstArgument() {
        underTest.setDotRadiusRange(0.49f, 1f)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeInvalidSecondArgument() {
        underTest.setDotRadiusRange(2f, 0f)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeInvalidBothArguments() {
        underTest.setDotRadiusRange(0.1f, -2f)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeFirstArgumentNAN() {
        underTest.setDotRadiusRange(java.lang.Float.NaN, 1f)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeSecondArgumentNAN() {
        underTest.setDotRadiusRange(1f, java.lang.Float.NaN)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeBothArgumentsNAN() {
        underTest.setDotRadiusRange(java.lang.Float.NaN, java.lang.Float.NaN)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetDotRadiusRangeMaxLessThanMin() {
        underTest.setDotRadiusRange(0.7f, 0.6f)
    }

    @Test
    fun setsDotRadiusRange() {
        underTest.setDotRadiusRange(0.5f, 0.6f)
        assertEquals(0.5f, underTest.minDotRadius, ASSERT_DELTA)
        assertEquals(0.6f, underTest.maxDotRadius, ASSERT_DELTA)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetLineThicknessInvalidArgument() {
        underTest.lineThickness = 0.99f
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetLineThicknessNAN() {
        underTest.lineThickness = java.lang.Float.NaN
    }

    @Test
    fun setsLineThickness() {
        underTest.lineThickness = 1f
        assertEquals(1f, underTest.lineThickness, ASSERT_DELTA)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetLineDistanceInvalidArgument() {
        underTest.lineDistance = java.lang.Float.NEGATIVE_INFINITY
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetLineDistanceNAN() {
        underTest.lineDistance = java.lang.Float.NaN
    }

    @Test
    fun setsLineDistance() {
        underTest.lineDistance = 0f
        assertEquals(0f, underTest.lineDistance, ASSERT_DELTA)
    }

    @Test(expected = IllegalArgumentException::class)
    fun crashesWhenSetNumDotsInvalidArgument() {
        underTest.numDots = -1
    }

    @Test
    fun setsNumDots() {
        underTest.numDots = 2
        assertEquals(2, underTest.numDots)
    }

    @Test
    fun setsLineColor() {
        underTest.lineColor = 2
        assertEquals(2, underTest.lineColor)
    }
}
